cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java

```diff
@@ -153,6 +168,71 @@ public void setOutput(String datasetName) {
     this.outputDatasetName = datasetName;
   }
 
+  @Override
+  public void setInput(String inputDatasetName, Dataset dataset) {
+    if (!(dataset instanceof BatchReadable) && !(dataset instanceof InputFormatProvider)) {
+      throw new IllegalArgumentException("Input dataset must be a BatchReadable or InputFormatProvider.");
+    }
+    // splits will be set by the MapReduceRuntimeService if they are not directly set by the program.
+    this.inputDatasetName = inputDatasetName;
+    this.inputDataset = dataset;
+  }
+
+  //TODO: update this to allow a BatchWritable once the DatasetOutputFormat can support taking an instance
+  //      and not just the name
+  @Override
+  public void setOutput(String outputDatasetName, Dataset dataset) {
+    if (!(dataset instanceof OutputFormatProvider)) {
+      throw new IllegalArgumentException("Output dataset must be an OutputFormatProvider.");
+    }
+    this.outputDatasetName = outputDatasetName;
+    this.outputDataset = dataset;
+  }
+
+  /**
+   * Get the output dataset for the job. If the dataset instance was set at runtime, that instance is returned.
+   * If the dataset name was set at runtime, an instance for that name is returned. If nothing was set at runtime, the
+   * output dataset from the program spec is used. If no output dataset was specified anywhere, a null is returned.
+   *
+   * @return Input dataset for the MapReduce job.
```

Comment:
this can vary from job to job so I will leave 'job' in there but fix the beginning of the sentence